# About Case Study

Have you ever struggled to analyze big datasets in Excel then, you should have considered Pandas, powerful programming and data analysis toolbox which allows you to manipulate million-row data points in milliseconds with high processing power and user productivity.

In this project, You will going to work with real dataset which is about marketing campaign. Targeted marketing, by definition, is the strategy of tailoring and personalizing online advertising according to data acquired on some intended audience.

In other words, instead of sending out ads to an incredibly large audience, a subset is selected based on their traits, interests, and preferences.

Dataset has fields like customers personal details like name, year of birth, martial status etc and also their recent purchase behaviour which includes last purchase_date, recency score, purchases so far etc

## Problem Statement

Analyze the given marketing campaign dataset to find interesting insights and relation among the user attributes which should lead to increase in the effectiveness of same. The analysis and takeaway will be consumed by marketing team to implement next marketing campaign.

In [2]:
import pandas as pd
import numpy as np
from datetime import date
import datetime as dt

Read the CSV data

In [3]:
## Read the data using pandas functions

dfp = pd.read_csv(r'C:\Users\USER\Downloads\relevel_python\customer_case_study\project_data.csv')
dfe = pd.read_csv(r'C:\Users\USER\Downloads\relevel_python\customer_case_study\edu_data.csv')
dfc_1 = pd.read_excel(r'C:\Users\USER\Downloads\relevel_python\customer_case_study\customer_data.xlsx', sheet_name='Data 1')
dfc_2 = pd.read_excel(r'C:\Users\USER\Downloads\relevel_python\customer_case_study\customer_data.xlsx', sheet_name='Data 2')

As we have two seperate files of customer_data. Let's concatenate them to make one.

In [4]:
## Concatenate or append both customer_data using concat function

dfc = pd.concat([dfc_1, dfc_2], ignore_index=True)

Let's go through each data table and skim through the fields

In [5]:
## Print the head of the project data

dfp.head(10)

customer_id purhcase_date  recency  online_purchases  store_purchases  \
0     20201701    09-04-2012       58                 8                4   
1     20201702    03-08-2014       38                 1                2   
2     20201703     8/21/2013       26                 8               10   
3     20201704    02-10-2014       26                 2                4   
4     20201705     1/19/2014       94                 5                6   
5     20201706    09-09-2013       16                 6               10   
6     20201707    11/13/2012       34                 7                7   
7     20201708    05-08-2013       32                 4                4   
8     20201709    06-06-2013       19                 3                2   
9     20201710     3/13/2014       68                 1                0   

   complaints  calls  intercoms  
0           0      3         11  
1           0      3         11  
2           0      3         11  
3           0      3         11  
4           0      3         11  
5           5      3         11  
6           0      3         11  
7           0      3         11  
8           3      3         11  
9           5      3         11

In [54]:
#checking the duplicate values
dfp.nunique()[0]-len(dfp.customer_id)

0

In [7]:
## Print the head of the education data

dfe.head(10)

educational_level educational_status_code
0        Graduation                      GD
1               PhD                      PH
2            Master                      MS
3             Basic                      BS
4       High School                      HS

In [8]:
## Print the head of the customer data

dfc.head(10)

cust_id  year_of_birth educational_status_code marital_status  \
0  20201733.0         1940.0                      GD        Married   
1  20202059.0         1943.0                     NaN        Married   
2  20202195.0         1944.0                      PH       Divorced   
3  20201983.0         1946.0                      GD   Relationship   
4  20202060.0         1948.0                      PH       Divorced   
5  20201723.0         1949.0                      PH        Widowed   
6  20201990.0         1949.0                      MS       Divorced   
7  20202199.0         1949.0                      PH   Relationship   
8  20201702.0         1950.0                      GD        Married   
9  20201725.0         1951.0                      GD       Divorced   

   annual_income              full_name  
0        40548.0            Abhinav Bal  
1        65073.0            Nimai Neela  
2        55614.0         Somatra Savant  
3        46276.0            Kamana Nath  
4        46681.0  Ninarika Chandrasekar  
5        58607.0          Naveen Badami  
6        35416.0          Kapila Kapoor  
7        41461.0           Sthir Sehgal  
8        46344.0       Advitiya Agarwal  
9        40689.0            Aarav Bahri

In [9]:
dfc.nunique()[0]-len(dfc.cust_id)  #there is one duplicate data in the data set

-1

In [10]:
dfc[dfc.duplicated('cust_id')==True]

cust_id  year_of_birth educational_status_code marital_status  \
318  20202110.0         1991.0                      GD        Married   

     annual_income        full_name  
318        50437.0  Tarangini Sodhi

Let's find out the size of the each data table

In [11]:
## Print the shape all data table

dfp.shape

(499, 8)

In [12]:
dfe.shape

(5, 2)

In [13]:
dfc.shape

(391, 6)

In [14]:
## Print the info of all data tables

dfp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 499 entries, 0 to 498
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customer_id       499 non-null    int64 
 1   purhcase_date     499 non-null    object
 2   recency           499 non-null    int64 
 3   online_purchases  499 non-null    int64 
 4   store_purchases   499 non-null    int64 
 5   complaints        499 non-null    int64 
 6   calls             499 non-null    int64 
 7   intercoms         499 non-null    int64 
dtypes: int64(7), object(1)
memory usage: 31.3+ KB


In [15]:
dfc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   cust_id                  391 non-null    float64
 1   year_of_birth            391 non-null    float64
 2   educational_status_code  363 non-null    object 
 3   marital_status           391 non-null    object 
 4   annual_income            373 non-null    float64
 5   full_name                391 non-null    object 
dtypes: float64(3), object(3)
memory usage: 18.5+ KB


Another interesting method is data.info() which gives us the number of data points and variables of the dataset. It also displays the data types. We can see that our dataset has 499 data points and 12 variables ranging from customers’ personal information to purchases, calls, intercoms, and complaints.

Let's perform a high level descriptive analysis using pandas describe method.

In [16]:
## Get the high level analysis done using describe function

dfp.describe()

customer_id     recency  online_purchases  store_purchases  \
count  4.990000e+02  499.000000        499.000000       499.000000   
mean   2.020195e+07   48.228457          3.969940         5.779559   
std    1.441932e+02   29.077101          2.813675         3.301378   
min    2.020170e+07    0.000000          0.000000         0.000000   
25%    2.020183e+07   23.000000          2.000000         3.000000   
50%    2.020195e+07   47.000000          3.000000         5.000000   
75%    2.020207e+07   74.000000          6.000000         8.000000   
max    2.020220e+07   99.000000         27.000000        13.000000   

       complaints       calls   intercoms  
count  499.000000  499.000000  499.000000  
mean     0.252505    3.370741   10.168337  
std      1.115953    4.327091    2.800552  
min      0.000000    0.000000    2.000000  
25%      0.000000    3.000000   11.000000  
50%      0.000000    3.000000   11.000000  
75%      0.000000    3.000000   11.000000  
max     11.000000   88.000000   44.000000

Descriptive Statistics
#1
The recency factor is based on the notion that the more recently a customer has made a purchase with a company, the more likely they will continue to keep the business and brand in mind for subsequent purchases.

As minimum recency score is 0 and maximum 99, data looks even distributed as 25% of the data has recency of 23, 50% of the data has recency of 47 etc. Also, As mean (i.e average) is close to median.

#2
Analysing purchase pattern, even though maximum puchase volume is 27 it looks like customer's aren't making much online purchase as 75% of distribution is till online purchase of volume 6

In [17]:
## Analyzing customer_data
dfc.describe(include="all")

## All values in project_data is numeric hence no point of using include="all"

cust_id  year_of_birth educational_status_code marital_status  \
count   3.910000e+02     391.000000                     363            391   
unique           NaN            NaN                       5              6   
top              NaN            NaN                      GD        Married   
freq             NaN            NaN                     195             94   
mean    2.020195e+07    1978.621483                     NaN            NaN   
std     1.461739e+02      13.844927                     NaN            NaN   
min     2.020170e+07    1940.000000                     NaN            NaN   
25%     2.020182e+07    1970.000000                     NaN            NaN   
50%     2.020195e+07    1982.000000                     NaN            NaN   
75%     2.020208e+07    1989.000000                     NaN            NaN   
max     2.020220e+07    2000.000000                     NaN            NaN   

        annual_income    full_name  
count      373.000000          391  
unique            NaN          391  
top               NaN  Abhinav Bal  
freq              NaN            1  
mean     51113.949062          NaN  
std      21152.035692          NaN  
min       2447.000000          NaN  
25%      34554.000000          NaN  
50%      50520.000000          NaN  
75%      68491.000000          NaN  
max     102692.000000          NaN

Descriptive Statistics
#1
Year of Birth should not be considered as numerical as its mean of 1978.36 doesn't make sense. So we can safely ignore it.

#2
Most of the customer we have has status Married.

In [18]:
## Check if any table has a missing value using isnull and sum function

dfp.isnull().sum()

customer_id         0
purhcase_date       0
recency             0
online_purchases    0
store_purchases     0
complaints          0
calls               0
intercoms           0
dtype: int64

In [19]:
dfc.isnull().sum()

cust_id                     0
year_of_birth               0
educational_status_code    28
marital_status              0
annual_income              18
full_name                   0
dtype: int64

In [20]:
## Print the data where annual_income is missing

dfc[dfc["annual_income"].isnull()]

cust_id  year_of_birth educational_status_code marital_status  \
10   20201749.0         1951.0                      GD        Married   
33   20201791.0         1957.0                      PH   Relationship   
34   20201792.0         1957.0                      GD        Married   
49   20202157.0         1957.0                      HS         Single   
56   20201744.0         1959.0                      PH       Divorced   
59   20201829.0         1961.0                      PH        Married   
130  20201793.0         1973.0                      MS   Relationship   
143  20201985.0         1975.0                      MS        Married   
171  20201834.0         1980.0                      GD   Relationship   
197  20201759.0         1982.0                      GD   Relationship   
211  20202005.0         1982.0                      GD   Relationship   
217  20201711.0         1983.0                      GD         Single   
259  20202105.0         1986.0                      PH       Divorced   
275  20201874.0         1988.0                      PH        Married   
283  20201772.0         1988.0                      HS         Single   
304  20202013.0         1990.0                      GD         Single   
342  20201779.0         1995.0                      HS   Relationship   
362  20202020.0         1997.0                      GD       Divorced   

     annual_income            full_name  
10             NaN         Adrika Bandi  
33             NaN         Ananya Bhatt  
34             NaN  Anaya Bhattacharyya  
49             NaN         Ramesh Saini  
56             NaN          Adita Balan  
59             NaN      Dharinija Chana  
130            NaN        Svara Shankar  
143            NaN            Kami Goel  
171            NaN       Dhruv Chandran  
197            NaN      Akanksha Barman  
211            NaN         Mayank Raman  
217            NaN        Avidosa Amble  
259            NaN           Pavan Raja  
275            NaN         Gayatri Toor  
283            NaN         Tanuja Sibal  
304            NaN        Minda Narayan  
342            NaN         Amrit Bhalla  
362            NaN            Mojil Roy

We know that in customer_data has missing annual income of the customer. Let's try to impute them. Although we can simply take average of all the customer's salary but we should not do that as different classes might have different salary for various reasons. For example a single individual might have less salary compared to married working couple. Hence let's take average salary based on the category and replace missing values.



In [21]:
## Check what all martial status available

dfc['marital_status'].unique()

array(['Married', 'Divorced', 'Relationship', 'Widowed', 'Single',
       'Widow'], dtype=object)

Let us consider we are interested to know the unique values of the marital status field then, we should select the column and apply the unique method. As shown below the variable marital status has 5 unique categories. However, we notice that widow and widowed are two two different naming for the same category so we can make it consistent through using replace method on the column values as displayed below.

In [22]:
## Replace one of the status using string's replace function

dfc['marital_status'] = dfc['marital_status'].replace('Widow', 'Widowed')

In [23]:
## Check martial status again

dfc['marital_status'].unique()

array(['Married', 'Divorced', 'Relationship', 'Widowed', 'Single'],
      dtype=object)

In [24]:
## Find out the average salary based on each martial_status. 
## Additionally, reset the index to convert into Df 

annual_income_avg = dfc.groupby("marital_status")["annual_income"].mean().reset_index()
annual_income_avg

marital_status  annual_income
0       Divorced   50672.338235
1        Married   49201.921348
2   Relationship   50609.129412
3         Single   53253.550000
4        Widowed   52524.568627

In [25]:
pd.isna(dfc['annual_income']).sum()

18

In [26]:
dfc[dfc['annual_income'].isnull()]

cust_id  year_of_birth educational_status_code marital_status  \
10   20201749.0         1951.0                      GD        Married   
33   20201791.0         1957.0                      PH   Relationship   
34   20201792.0         1957.0                      GD        Married   
49   20202157.0         1957.0                      HS         Single   
56   20201744.0         1959.0                      PH       Divorced   
59   20201829.0         1961.0                      PH        Married   
130  20201793.0         1973.0                      MS   Relationship   
143  20201985.0         1975.0                      MS        Married   
171  20201834.0         1980.0                      GD   Relationship   
197  20201759.0         1982.0                      GD   Relationship   
211  20202005.0         1982.0                      GD   Relationship   
217  20201711.0         1983.0                      GD         Single   
259  20202105.0         1986.0                      PH       Divorced   
275  20201874.0         1988.0                      PH        Married   
283  20201772.0         1988.0                      HS         Single   
304  20202013.0         1990.0                      GD         Single   
342  20201779.0         1995.0                      HS   Relationship   
362  20202020.0         1997.0                      GD       Divorced   

     annual_income            full_name  
10             NaN         Adrika Bandi  
33             NaN         Ananya Bhatt  
34             NaN  Anaya Bhattacharyya  
49             NaN         Ramesh Saini  
56             NaN          Adita Balan  
59             NaN      Dharinija Chana  
130            NaN        Svara Shankar  
143            NaN            Kami Goel  
171            NaN       Dhruv Chandran  
197            NaN      Akanksha Barman  
211            NaN         Mayank Raman  
217            NaN        Avidosa Amble  
259            NaN           Pavan Raja  
275            NaN         Gayatri Toor  
283            NaN         Tanuja Sibal  
304            NaN        Minda Narayan  
342            NaN         Amrit Bhalla  
362            NaN            Mojil Roy

In [27]:
## Write an function to replace missing annual income value with its average (based on martial status)
def annual_income_avg_def (a_i, ms) :
  lst = []
  j = a_i[a_i['marital_status']==ms].annual_income.mean()
  for i in a_i.annual_income:         ## Important Condition
    if(pd.isna(i)):
      lst.append(j)
    else:
      lst.append(i)
  return lst
annual_income_avg_def(dfc,'Married')

[40548.0,
 65073.0,
 55614.0,
 46276.0,
 46681.0,
 58607.0,
 35416.0,
 41461.0,
 46344.0,
 40689.0,
 49201.92134831461,
 53790.0,
 30096.0,
 73448.0,
 38872.0,
 77972.0,
 75278.0,
 55951.0,
 58821.0,
 31615.0,
 33322.0,
 59354.0,
 54348.0,
 64500.0,
 72550.0,
 85503.0,
 26527.0,
 23018.0,
 70626.0,
 50898.0,
 62503.0,
 21994.0,
 50388.0,
 49201.92134831461,
 49201.92134831461,
 88193.0,
 101970.0,
 60199.0,
 30523.0,
 70356.0,
 44159.0,
 37717.0,
 14849.0,
 29999.0,
 24072.0,
 71113.0,
 78618.0,
 18492.0,
 48686.0,
 49201.92134831461,
 58723.0,
 74293.0,
 40737.0,
 50520.0,
 70991.0,
 63033.0,
 49201.92134831461,
 82800.0,
 62204.0,
 49201.92134831461,
 59432.0,
 70046.0,
 33378.0,
 66313.0,
 80124.0,
 79143.0,
 78825.0,
 58308.0,
 71613.0,
 84618.0,
 83003.0,
 70287.0,
 38179.0,
 79946.0,
 62000.0,
 19510.0,
 80317.0,
 51479.0,
 93027.0,
 70777.0,
 54178.0,
 41728.0,
 69674.0,
 55954.0,
 75693.0,
 18589.0,
 38361.0,
 55212.0,
 68462.0,
 62187.0,
 14045.0,
 37760.0,
 36138.0,
 37054.0,

In [29]:
# Function Testing
ans = annual_income_avg_def(dfc, "Married")
type(ans)

list

In [30]:
left = dfc[dfc['annual_income'].isnull()]
t1 = pd.merge(left, annual_income_avg, how = 'left', on = 'marital_status')
t1

cust_id  year_of_birth educational_status_code marital_status  \
0   20201749.0         1951.0                      GD        Married   
1   20201791.0         1957.0                      PH   Relationship   
2   20201792.0         1957.0                      GD        Married   
3   20202157.0         1957.0                      HS         Single   
4   20201744.0         1959.0                      PH       Divorced   
5   20201829.0         1961.0                      PH        Married   
6   20201793.0         1973.0                      MS   Relationship   
7   20201985.0         1975.0                      MS        Married   
8   20201834.0         1980.0                      GD   Relationship   
9   20201759.0         1982.0                      GD   Relationship   
10  20202005.0         1982.0                      GD   Relationship   
11  20201711.0         1983.0                      GD         Single   
12  20202105.0         1986.0                      PH       Divorced   
13  20201874.0         1988.0                      PH        Married   
14  20201772.0         1988.0                      HS         Single   
15  20202013.0         1990.0                      GD         Single   
16  20201779.0         1995.0                      HS   Relationship   
17  20202020.0         1997.0                      GD       Divorced   

    annual_income_x            full_name  annual_income_y  
0               NaN         Adrika Bandi     49201.921348  
1               NaN         Ananya Bhatt     50609.129412  
2               NaN  Anaya Bhattacharyya     49201.921348  
3               NaN         Ramesh Saini     53253.550000  
4               NaN          Adita Balan     50672.338235  
5               NaN      Dharinija Chana     49201.921348  
6               NaN        Svara Shankar     50609.129412  
7               NaN            Kami Goel     49201.921348  
8               NaN       Dhruv Chandran     50609.129412  
9               NaN      Akanksha Barman     50609.129412  
10              NaN         Mayank Raman     50609.129412  
11              NaN        Avidosa Amble     53253.550000  
12              NaN           Pavan Raja     50672.338235  
13              NaN         Gayatri Toor     49201.921348  
14              NaN         Tanuja Sibal     53253.550000  
15              NaN        Minda Narayan     53253.550000  
16              NaN         Amrit Bhalla     50609.129412  
17              NaN            Mojil Roy     50672.338235

In [31]:
t1 = t1.drop('annual_income_x', axis = 1)
t1

cust_id  year_of_birth educational_status_code marital_status  \
0   20201749.0         1951.0                      GD        Married   
1   20201791.0         1957.0                      PH   Relationship   
2   20201792.0         1957.0                      GD        Married   
3   20202157.0         1957.0                      HS         Single   
4   20201744.0         1959.0                      PH       Divorced   
5   20201829.0         1961.0                      PH        Married   
6   20201793.0         1973.0                      MS   Relationship   
7   20201985.0         1975.0                      MS        Married   
8   20201834.0         1980.0                      GD   Relationship   
9   20201759.0         1982.0                      GD   Relationship   
10  20202005.0         1982.0                      GD   Relationship   
11  20201711.0         1983.0                      GD         Single   
12  20202105.0         1986.0                      PH       Divorced   
13  20201874.0         1988.0                      PH        Married   
14  20201772.0         1988.0                      HS         Single   
15  20202013.0         1990.0                      GD         Single   
16  20201779.0         1995.0                      HS   Relationship   
17  20202020.0         1997.0                      GD       Divorced   

              full_name  annual_income_y  
0          Adrika Bandi     49201.921348  
1          Ananya Bhatt     50609.129412  
2   Anaya Bhattacharyya     49201.921348  
3          Ramesh Saini     53253.550000  
4           Adita Balan     50672.338235  
5       Dharinija Chana     49201.921348  
6         Svara Shankar     50609.129412  
7             Kami Goel     49201.921348  
8        Dhruv Chandran     50609.129412  
9       Akanksha Barman     50609.129412  
10         Mayank Raman     50609.129412  
11        Avidosa Amble     53253.550000  
12           Pavan Raja     50672.338235  
13         Gayatri Toor     49201.921348  
14         Tanuja Sibal     53253.550000  
15        Minda Narayan     53253.550000  
16         Amrit Bhalla     50609.129412  
17            Mojil Roy     50672.338235

In [32]:
t1 = t1.rename(columns = {'annual_income_y': 'annual_income'})
t1

cust_id  year_of_birth educational_status_code marital_status  \
0   20201749.0         1951.0                      GD        Married   
1   20201791.0         1957.0                      PH   Relationship   
2   20201792.0         1957.0                      GD        Married   
3   20202157.0         1957.0                      HS         Single   
4   20201744.0         1959.0                      PH       Divorced   
5   20201829.0         1961.0                      PH        Married   
6   20201793.0         1973.0                      MS   Relationship   
7   20201985.0         1975.0                      MS        Married   
8   20201834.0         1980.0                      GD   Relationship   
9   20201759.0         1982.0                      GD   Relationship   
10  20202005.0         1982.0                      GD   Relationship   
11  20201711.0         1983.0                      GD         Single   
12  20202105.0         1986.0                      PH       Divorced   
13  20201874.0         1988.0                      PH        Married   
14  20201772.0         1988.0                      HS         Single   
15  20202013.0         1990.0                      GD         Single   
16  20201779.0         1995.0                      HS   Relationship   
17  20202020.0         1997.0                      GD       Divorced   

              full_name  annual_income  
0          Adrika Bandi   49201.921348  
1          Ananya Bhatt   50609.129412  
2   Anaya Bhattacharyya   49201.921348  
3          Ramesh Saini   53253.550000  
4           Adita Balan   50672.338235  
5       Dharinija Chana   49201.921348  
6         Svara Shankar   50609.129412  
7             Kami Goel   49201.921348  
8        Dhruv Chandran   50609.129412  
9       Akanksha Barman   50609.129412  
10         Mayank Raman   50609.129412  
11        Avidosa Amble   53253.550000  
12           Pavan Raja   50672.338235  
13         Gayatri Toor   49201.921348  
14         Tanuja Sibal   53253.550000  
15        Minda Narayan   53253.550000  
16         Amrit Bhalla   50609.129412  
17            Mojil Roy   50672.338235

In [33]:
dfc[~dfc['annual_income'].isnull()]

cust_id  year_of_birth educational_status_code marital_status  \
0    20201733.0         1940.0                      GD        Married   
1    20202059.0         1943.0                     NaN        Married   
2    20202195.0         1944.0                      PH       Divorced   
3    20201983.0         1946.0                      GD   Relationship   
4    20202060.0         1948.0                      PH       Divorced   
..          ...            ...                     ...            ...   
386  20201886.0         2000.0                      GD         Single   
387  20202023.0         2000.0                      PH   Relationship   
388  20202089.0         2000.0                      GD   Relationship   
389  20202119.0         2000.0                      GD         Single   
390  20202153.0         2000.0                      MS         Single   

     annual_income              full_name  
0          40548.0            Abhinav Bal  
1          65073.0            Nimai Neela  
2          55614.0         Somatra Savant  
3          46276.0            Kamana Nath  
4          46681.0  Ninarika Chandrasekar  
..             ...                    ...  
386        25271.0          Gurnam Varkey  
387        26476.0           Mrinal Nigam  
388        38823.0           Oormi Srivas  
389        91065.0             Taru Soman  
390        36230.0          Tejas Sridhar  

[373 rows x 6 columns]

In [34]:
dfc = pd.concat([t1, dfc[~dfc['annual_income'].isnull()]])
dfc

cust_id  year_of_birth educational_status_code marital_status  \
0    20201749.0         1951.0                      GD        Married   
1    20201791.0         1957.0                      PH   Relationship   
2    20201792.0         1957.0                      GD        Married   
3    20202157.0         1957.0                      HS         Single   
4    20201744.0         1959.0                      PH       Divorced   
..          ...            ...                     ...            ...   
386  20201886.0         2000.0                      GD         Single   
387  20202023.0         2000.0                      PH   Relationship   
388  20202089.0         2000.0                      GD   Relationship   
389  20202119.0         2000.0                      GD         Single   
390  20202153.0         2000.0                      MS         Single   

               full_name  annual_income  
0           Adrika Bandi   49201.921348  
1           Ananya Bhatt   50609.129412  
2    Anaya Bhattacharyya   49201.921348  
3           Ramesh Saini   53253.550000  
4            Adita Balan   50672.338235  
..                   ...            ...  
386        Gurnam Varkey   25271.000000  
387         Mrinal Nigam   26476.000000  
388         Oormi Srivas   38823.000000  
389           Taru Soman   91065.000000  
390        Tejas Sridhar   36230.000000  

[391 rows x 6 columns]

In [35]:
## Checking annual_income column for null values
dfc['annual_income'].isnull().sum()

0

In [36]:
dfc.isnull().sum()

cust_id                     0
year_of_birth               0
educational_status_code    28
marital_status              0
full_name                   0
annual_income               0
dtype: int64

As we need to do thorough analysis we will get  data from various tables possible. Hence let's merge them based on primary keys.

In [37]:
## Left join Project and Customer table using merge function
## Make sure to keep project_data as left table and customer_data as right table 
## So that with left join we don't lose the missing customer data information
## Save it in the temporary DF as we need to join one more table

df_temp = pd.merge(dfp, dfc, left_on='customer_id', right_on='cust_id', how='left')
df_temp

customer_id purhcase_date  recency  online_purchases  store_purchases  \
0       20201701    09-04-2012       58                 8                4   
1       20201702    03-08-2014       38                 1                2   
2       20201703     8/21/2013       26                 8               10   
3       20201704    02-10-2014       26                 2                4   
4       20201705     1/19/2014       94                 5                6   
..           ...           ...      ...               ...              ...   
495     20202195    11/27/2013       85                 9                6   
496     20202196     4/13/2013       88                 5               11   
497     20202197    04-05-2014       22                 2                3   
498     20202198     4/21/2013       83                 4                4   
499     20202199     5/22/2014       63                 6               11   

     complaints  calls  intercoms     cust_id  year_of_birth  \
0             0      3         11  20201701.0         1982.0   
1             0      3         11  20201702.0         1950.0   
2             0      3         11  20201703.0         1965.0   
3             0      3         11         NaN            NaN   
4             0      3         11  20201705.0         1981.0   
..          ...    ...        ...         ...            ...   
495           0      3         11  20202195.0         1944.0   
496           0      3         11  20202196.0         1962.0   
497           0      3         11  20202197.0         1978.0   
498           0      6         11  20202198.0         1971.0   
499           0      6         11  20202199.0         1949.0   

    educational_status_code marital_status         full_name  annual_income  
0                        GD         Single     Aditi Acharya        58138.0  
1                        GD        Married  Advitiya Agarwal        46344.0  
2                        GD       Divorced    Alagesan Agate        71613.0  
3                       NaN            NaN               NaN            NaN  
4                        PH        Widowed   Amrish Aggarwal        58293.0  
..                      ...            ...               ...            ...  
495                      PH       Divorced    Somatra Savant        55614.0  
496                      MS       Divorced      Sona Sawhney        59432.0  
497                      GD       Divorced      Tamal Shenoy        55563.0  
498                      PH   Relationship     Souvik Saxena        43624.0  
499                      PH   Relationship      Sthir Sehgal        41461.0  

[500 rows x 14 columns]

In [38]:
df_temp.shape

(500, 14)

In [39]:
df_temp.isnull().sum()

customer_id                  0
purhcase_date                0
recency                      0
online_purchases             0
store_purchases              0
complaints                   0
calls                        0
intercoms                    0
cust_id                    109
year_of_birth              109
educational_status_code    137
marital_status             109
full_name                  109
annual_income              109
dtype: int64

In [40]:
## Join temporary joined table with education data table

df = pd.merge(df_temp, dfe, left_on='educational_status_code', right_on='educational_status_code', how='left')
df

customer_id purhcase_date  recency  online_purchases  store_purchases  \
0       20201701    09-04-2012       58                 8                4   
1       20201702    03-08-2014       38                 1                2   
2       20201703     8/21/2013       26                 8               10   
3       20201704    02-10-2014       26                 2                4   
4       20201705     1/19/2014       94                 5                6   
..           ...           ...      ...               ...              ...   
495     20202195    11/27/2013       85                 9                6   
496     20202196     4/13/2013       88                 5               11   
497     20202197    04-05-2014       22                 2                3   
498     20202198     4/21/2013       83                 4                4   
499     20202199     5/22/2014       63                 6               11   

     complaints  calls  intercoms     cust_id  year_of_birth  \
0             0      3         11  20201701.0         1982.0   
1             0      3         11  20201702.0         1950.0   
2             0      3         11  20201703.0         1965.0   
3             0      3         11         NaN            NaN   
4             0      3         11  20201705.0         1981.0   
..          ...    ...        ...         ...            ...   
495           0      3         11  20202195.0         1944.0   
496           0      3         11  20202196.0         1962.0   
497           0      3         11  20202197.0         1978.0   
498           0      6         11  20202198.0         1971.0   
499           0      6         11  20202199.0         1949.0   

    educational_status_code marital_status         full_name  annual_income  \
0                        GD         Single     Aditi Acharya        58138.0   
1                        GD        Married  Advitiya Agarwal        46344.0   
2                        GD       Divorced    Alagesan Agate        71613.0   
3                       NaN            NaN               NaN            NaN   
4                        PH        Widowed   Amrish Aggarwal        58293.0   
..                      ...            ...               ...            ...   
495                      PH       Divorced    Somatra Savant        55614.0   
496                      MS       Divorced      Sona Sawhney        59432.0   
497                      GD       Divorced      Tamal Shenoy        55563.0   
498                      PH   Relationship     Souvik Saxena        43624.0   
499                      PH   Relationship      Sthir Sehgal        41461.0   

    educational_level  
0          Graduation  
1          Graduation  
2          Graduation  
3                 NaN  
4                 PhD  
..                ...  
495               PhD  
496            Master  
497        Graduation  
498               PhD  
499               PhD  

[500 rows x 15 columns]

Do we have any missing data?

Let's check.

In [41]:
df.isnull().sum()

customer_id                  0
purhcase_date                0
recency                      0
online_purchases             0
store_purchases              0
complaints                   0
calls                        0
intercoms                    0
cust_id                    109
year_of_birth              109
educational_status_code    137
marital_status             109
full_name                  109
annual_income              109
educational_level          137
dtype: int64

It looks like we don't have customer information of 109 customers as we see 109 cust_id missing in compare to customer_id.

Task - Try joining the table with inner join.

In [42]:
## Filter the educational level and find the number of records for each
## Additionally sort the value in descending order

df['educational_level'].value_counts().sort_values(ascending=False)

Graduation     195
PhD             80
Master          54
High School     28
Basic            6
Name: educational_level, dtype: int64

In [43]:
#another way using groupby
#df.groupby(by="educational_level")['educational_level'].count()

Print only customer details of the customer and save it in different DataFrame.

DataFrame should contain 'full_name','year_of_birth', 'educational_level', 'annual_income' columns.

In [44]:
##create a derived table from main table as per the columns mentioned above

edu_details=df[['full_name','year_of_birth', 'educational_level', 'annual_income']]
edu_details

full_name  year_of_birth educational_level  annual_income
0       Aditi Acharya         1982.0        Graduation        58138.0
1    Advitiya Agarwal         1950.0        Graduation        46344.0
2      Alagesan Agate         1965.0        Graduation        71613.0
3                 NaN            NaN               NaN            NaN
4     Amrish Aggarwal         1981.0               PhD        58293.0
..                ...            ...               ...            ...
495    Somatra Savant         1944.0               PhD        55614.0
496      Sona Sawhney         1962.0            Master        59432.0
497      Tamal Shenoy         1978.0        Graduation        55563.0
498     Souvik Saxena         1971.0               PhD        43624.0
499      Sthir Sehgal         1949.0               PhD        41461.0

[500 rows x 4 columns]

Display those customers who have done Masters.

In [45]:
df[df['educational_level']=='Master']

customer_id purhcase_date  recency  online_purchases  store_purchases  \
5       20201706    09-09-2013       16                 6               10   
13      20201714    11/15/2013       53                 6                5   
18      20201719     3/28/2013       91                11                9   
30      20201731     5/22/2014       34                 3                3   
34      20201735     2/20/2013        4                 3                9   
44      20201745    10-02-2013       19                 2                3   
51      20201752    08-11-2012        2                 6               13   
60      20201761    07-04-2013       12                 3               11   
70      20201771    09-11-2012        1                 4                6   
76      20201777     8/27/2012       34                 7                5   
86      20201787     7/16/2013       48                 3                4   
92      20201793    11/23/2013       87                 2                8   
93      20201794     3/23/2014       69                 1                3   
120     20201821     6/20/2013       65                 6               10   
122     20201823    09-04-2013       63                 1                3   
127     20201828    06-08-2014       51                 2                4   
131     20201832    03-10-2013       42                11                6   
140     20201841    10/19/2012       46                 5               12   
149     20201850     6/21/2014       65                 2                4   
196     20201897     5/13/2014       80                 7               12   
205     20201906    11-10-2012        5                 0                2   
213     20201914     1/21/2013       21                 7                8   
217     20201918    11/15/2013       81                 8                6   
236     20201937    12-04-2013       46                 1                3   
249     20201950    10/29/2012       57                 9                6   
262     20201963     6/17/2014       85                 4                6   
265     20201966     3/14/2013       30                 3                3   
284     20201985     4/17/2014       18                 2                3   
287     20201988    01-08-2013       88                 6               12   
289     20201990     6/13/2013       62                 7                5   
295     20201996     8/23/2012       62                 3               11   
297     20201998     1/13/2013       43                 3                3   
309     20202010     4/27/2013       12                 4                4   
313     20202014    02-06-2013       42                 6               10   
372     20202073     8/29/2013       52                 1                2   
375     20202076     8/18/2012       86                 3                2   
384     20202085     3/30/2014       33                 0                2   
386     20202087    10/16/2013       11                 7                4   
391     20202092    12-09-2013        0                 3                6   
394     20202095     2/25/2014       80                 5               12   
420     20202120    03-05-2013        4                10                8   
423     20202123    06-05-2014       75                 1                2   
424     20202124    06-04-2013       26                 3                8   
436     20202136     6/21/2013       49                 1                3   
438     20202138    09-07-2013       14                 9                5   
447     20202147    10-09-2013       92                 5                4   
453     20202153    10/17/2013       17                 2                4   
458     20202158    04-01-2013       71                 4                5   
478     20202178     5/23/2014       19                 7                8   
481     20202181    12/30/2013       60                 2                2   
482     20202182    08-11-2

Display the top 10 records of the customer based on their recency.

Record should display 'full_name','educational_level', 'annual_income' columns

In [46]:
## Display the top 10 records of the customer based on their recency using loc function
## Additionally sort them using sort_values function
df.loc[0:9,['full_name','educational_level', 'annual_income','recency']].sort_values(by='recency',ascending=False).reset_index()

index            full_name educational_level  annual_income  recency
0      4      Amrish Aggarwal               PhD        58293.0       94
1      9                  NaN               NaN            NaN       68
2      0        Aditi Acharya        Graduation        58138.0       58
3      1     Advitiya Agarwal        Graduation        46344.0       38
4      6  Asvathama Ahluwalia        Graduation        55635.0       34
5      7                  NaN               NaN            NaN       32
6      2       Alagesan Agate        Graduation        71613.0       26
7      3                  NaN               NaN            NaN       26
8      8        Avantas Ahuja               PhD        30351.0       19
9      5  Aprativirya Agrawal            Master        62000.0       16

In [47]:
##without having revency in column
df.nlargest(10,'recency')[['full_name','educational_level','annual_income']].reset_index()

index         full_name educational_level  annual_income
0     38       Adarsh Bala        Graduation        42429.0
1    192        Hansa Vasa       High School        36640.0
2    208       Hiral Kunda        Graduation        44794.0
3    445               NaN               NaN            NaN
4    492      Sneha Sarraf        Graduation        46310.0
5    164               NaN               NaN            NaN
6    191  Hajari Varughese               PhD        66973.0
7    240               NaN               NaN            NaN
8    241     Javed Mallick        Graduation        64961.0
9    257               NaN               NaN            NaN

In [48]:
##top 10 based on annual income

df.nlargest(10,'annual_income').reset_index()

index  customer_id purhcase_date  recency  online_purchases  \
0    252     20201953    04-05-2013        5                 6   
1    203     20201904    11-02-2012       54                 7   
2    124     20201825    03-12-2013       69                 6   
3    425     20202125     4/13/2013       77                 7   
4    140     20201841    10/19/2012       46                 5   
5    419     20202119     2/22/2013       33                 7   
6    120     20201821     6/20/2013       65                 6   
7    198     20201899     1/17/2013       20                 5   
8    407     20202108     5/30/2014        1                 5   
9    142     20201843     4/19/2013       73                 5   

   store_purchases  complaints  calls  intercoms     cust_id  year_of_birth  \
0               13           0      3         11  20201953.0         1997.0   
1               10           0      4         11  20201904.0         1989.0   
2               13           0      3         11  20201825.0         1957.0   
3                5           0      3         11  20202125.0         1967.0   
4               12           0      3          2  20201841.0         1987.0   
5                9           0      3         11  20202119.0         2000.0   
6               10           0      5         11  20201821.0         1957.0   
7                5           0      3          2  20201899.0         1984.0   
8                5           0      8         11  20202108.0         1989.0   
9               10           0      3         11  20201843.0         1989.0   

  educational_status_code marital_status            full_name  annual_income  \
0                      GD        Widowed            Jute Mani       102692.0   
1                      PH         Single          Himani Vyas       102160.0   
2                      GD         Single          Devi Chadha       101970.0   
3                      PH        Married    Prita Ranganathan        93027.0   
4                      MS         Single    Divya Subramaniam        92859.0   
5                      GD         Single           Taru Soman        91065.0   
6                      MS   Relationship           Deven Chad        88193.0   
7                      GD        Widowed           Harsha Vig        86718.0   
8                      PH   Relationship    Piroj Rajagopalan        86164.0   
9                      GD       Divorced  Divyesh Subramanian        86111.0   

  educational_level  
0        Graduation  
1               PhD  
2        Graduation  
3               PhD  
4            Master  
5        Graduation  
6            Master  
7        Graduation  
8               PhD  
9        Graduation

In [49]:
## doing the same using .iloc function
## since .iloc can only be accessed by the column  index, we need to find index number of required coulmns 
print(list(df.columns).index('full_name'))
print(list(df.columns).index('educational_level'))
print(list(df.columns).index('annual_income'))


12
14
13


In [50]:
df.iloc[:,[13,14,12]].sort_values(by='annual_income',ascending=False).reset_index().head(10)

index  annual_income educational_level            full_name
0    252       102692.0        Graduation            Jute Mani
1    203       102160.0               PhD          Himani Vyas
2    124       101970.0        Graduation          Devi Chadha
3    425        93027.0               PhD    Prita Ranganathan
4    140        92859.0            Master    Divya Subramaniam
5    419        91065.0        Graduation           Taru Soman
6    120        88193.0            Master           Deven Chad
7    198        86718.0        Graduation           Harsha Vig
8    407        86164.0               PhD    Piroj Rajagopalan
9    142        86111.0        Graduation  Divyesh Subramanian

Although we don't have exact birthdate of an customer but let's try to find age of the customer from birth year.

In [51]:
# Find out the age of customers based on the current year

today=date.today()
year=today.year
df['age']=year-df['year_of_birth']
df['age']

0      41.0
1      73.0
2      58.0
3       NaN
4      42.0
       ... 
495    79.0
496    61.0
497    45.0
498    52.0
499    74.0
Name: age, Length: 500, dtype: float64

In [52]:
##checking the age column in dataframe
df

customer_id purhcase_date  recency  online_purchases  store_purchases  \
0       20201701    09-04-2012       58                 8                4   
1       20201702    03-08-2014       38                 1                2   
2       20201703     8/21/2013       26                 8               10   
3       20201704    02-10-2014       26                 2                4   
4       20201705     1/19/2014       94                 5                6   
..           ...           ...      ...               ...              ...   
495     20202195    11/27/2013       85                 9                6   
496     20202196     4/13/2013       88                 5               11   
497     20202197    04-05-2014       22                 2                3   
498     20202198     4/21/2013       83                 4                4   
499     20202199     5/22/2014       63                 6               11   

     complaints  calls  intercoms     cust_id  year_of_birth  \
0             0      3         11  20201701.0         1982.0   
1             0      3         11  20201702.0         1950.0   
2             0      3         11  20201703.0         1965.0   
3             0      3         11         NaN            NaN   
4             0      3         11  20201705.0         1981.0   
..          ...    ...        ...         ...            ...   
495           0      3         11  20202195.0         1944.0   
496           0      3         11  20202196.0         1962.0   
497           0      3         11  20202197.0         1978.0   
498           0      6         11  20202198.0         1971.0   
499           0      6         11  20202199.0         1949.0   

    educational_status_code marital_status         full_name  annual_income  \
0                        GD         Single     Aditi Acharya        58138.0   
1                        GD        Married  Advitiya Agarwal        46344.0   
2                        GD       Divorced    Alagesan Agate        71613.0   
3                       NaN            NaN               NaN            NaN   
4                        PH        Widowed   Amrish Aggarwal        58293.0   
..                      ...            ...               ...            ...   
495                      PH       Divorced    Somatra Savant        55614.0   
496                      MS       Divorced      Sona Sawhney        59432.0   
497                      GD       Divorced      Tamal Shenoy        55563.0   
498                      PH   Relationship     Souvik Saxena        43624.0   
499                      PH   Relationship      Sthir Sehgal        41461.0   

    educational_level   age  
0          Graduation  41.0  
1          Graduation  73.0  
2          Graduation  58.0  
3                 NaN   NaN  
4                 PhD  42.0  
..                ...   ...  
495               PhD  79.0  
496            Master  61.0  
497        Graduation  45.0  
498               PhD  52.0  
499               PhD  74.0  

[500 rows x 16 columns]

In [53]:
pd.pivot_table(df,values=['age','annual_income'],index='marital_status',aggfunc='mean')

age  annual_income
marital_status                          
Divorced        45.845070   50672.338235
Married         45.212766   49201.921348
Relationship    46.032967   50609.129412
Single          41.523810   53253.550000
Widowed         42.549020   52524.568627

Let's find out the average age and annual income of the customers whoes martial status is either 'Widowed' or 'Divorced'. Based on the data special scheme can be planned for them.

In [54]:
## Filter the data of Widowed and Divorced martial status and find out their average age and annual income using mean function
df.loc[df['marital_status'].isin(['Widowed','Divorced'])][['age','annual_income']].mean()

age                 44.467213
annual_income    51446.631268
dtype: float64

Filter the number of customers with an income higher than 75,000 and with a master’s degree.

In [55]:
# Filter the number of customers with an income higher than 75,000 and with a master’s degree using loc function
df.loc[(df.annual_income>75000)&(df.educational_level=='Master')]

customer_id purhcase_date  recency  online_purchases  store_purchases  \
18      20201719     3/28/2013       91                11                9   
51      20201752    08-11-2012        2                 6               13   
60      20201761    07-04-2013       12                 3               11   
76      20201777     8/27/2012       34                 7                5   
120     20201821     6/20/2013       65                 6               10   
140     20201841    10/19/2012       46                 5               12   
217     20201918    11/15/2013       81                 8                6   
424     20202124    06-04-2013       26                 3                8   
436     20202136     6/21/2013       49                 1                3   

     complaints  calls  intercoms     cust_id  year_of_birth  \
18            4      3         11  20201719.0         1980.0   
51            0      3         11  20201752.0         1964.0   
60            0      3         11  20201761.0         1982.0   
76            0      3         11  20201777.0         1993.0   
120           0      5         11  20201821.0         1957.0   
140           0      3          2  20201841.0         1987.0   
217           0      3         11  20201918.0         1985.0   
424           0      3         11  20202124.0         1973.0   
436           0      0         11  20202136.0         1983.0   

    educational_status_code marital_status          full_name  annual_income  \
18                       MS         Single      Hardeep Atwal        76995.0   
51                       MS         Single       Agrata Banik        79143.0   
60                       MS         Single        Akhil Bassi        75777.0   
76                       MS        Married      Amithi Bhagat        75251.0   
120                      MS   Relationship         Deven Chad        88193.0   
140                      MS         Single  Divya Subramaniam        92859.0   
217                      MS        Widowed        Ishan Lalla        83790.0   
424                      MS   Relationship   Premala Randhawa        82584.0   
436                      MS   Relationship       Rachna Raval        82634.0   

    educational_level   age  
18             Master  43.0  
51             Master  59.0  
60             Master  41.0  
76             Master  30.0  
120            Master  66.0  
140            Master  36.0  
217            Master  38.0  
424            Master  50.0  
436            Master  40.0

In [56]:
## Extend the previous code to count the number of records i.e customer_id
df.loc[(df.annual_income>75000)&(df.educational_level=='Master'),'customer_id'].count()

9

In [60]:
# Create a new variable which is the sum of all purchases performed by customers
df['total_purchase']=df.online_purchases+df.store_purchases
df['total_purchase']

0      12
1       3
2      18
3       6
4      11
       ..
495    15
496    16
497     5
498     8
499    17
Name: total_purchase, Length: 500, dtype: int64

Find the maximum and minimum annual income of the customer.

In [61]:
## Find the maximum and minimum annual income of the customer using aggregation function

print(df['annual_income'].max())
print(df['annual_income'].min())

102692.0
2447.0


In [70]:
## Write an function to segregate annual income as per following
## Greater than 100000: High
## 50000< Salary < 100000: Medium
## Anything below 50000: Low

def income_category(annual_income):
    if (annual_income>=100000):
        return 'High'
    elif (50000<=annual_income<100000):
        return 'Medium'
    else:
        return 'Low'

In [71]:
# Create an income category (low, medium, high) based on the income variable using apply and lambda function

df['income_category']=df.apply(lambda x:income_category(x['annual_income']),axis=1)  #assign the categories based on income
df[['annual_income','income_category']]

annual_income income_category
0          58138.0          Medium
1          46344.0             Low
2          71613.0          Medium
3              NaN             Low
4          58293.0          Medium
..             ...             ...
495        55614.0          Medium
496        59432.0          Medium
497        55563.0          Medium
498        43624.0             Low
499        41461.0             Low

[500 rows x 2 columns]

Find out the number of customer belong to each income category.

In [72]:
## Find out the number of customer belong to each income category using group by
df.groupby('income_category').count()['customer_id']

income_category
High        3
Low       295
Medium    202
Name: customer_id, dtype: int64

We see very few customer who are in high income category. Display their name, annual income, education status and age.

In [74]:
## Display their name, annual income, education status and age using normal filtering
df[df['income_category']=='High'][['full_name','annual_income','educational_level','age','total_purchase']]

full_name  annual_income educational_level   age  total_purchase
124  Devi Chadha       101970.0        Graduation  66.0              19
203  Himani Vyas       102160.0               PhD  34.0              17
252    Jute Mani       102692.0        Graduation  26.0              19

Also, we see that we have more customers from low income category then medium income category. It will interesting to find the purchase pattern of both types of customer based on their income, age.

In [75]:
## Find out the list of columns of merged Dataframe
df.columns

Index(['customer_id', 'purhcase_date', 'recency', 'online_purchases',
       'store_purchases', 'complaints', 'calls', 'intercoms', 'cust_id',
       'year_of_birth', 'educational_status_code', 'marital_status',
       'full_name', 'annual_income', 'educational_level', 'age',
       'total_purchase', 'income_category'],
      dtype='object')

In [78]:
#apply groupby to find the mean of income, recency, number of web and store purchases by educational group
aggregate_view=pd.DataFrame(df.groupby(by='educational_level')[['annual_income','recency','online_purchases','store_purchases']].mean()).reset_index()
aggregate_view.sort_values(by=['recency'],ascending=False)

educational_level  annual_income    recency  online_purchases  \
2       High School   43883.508193  56.500000          3.357143   
0             Basic   18058.833333  49.166667          1.500000   
4               PhD   55086.583107  48.925000          4.637500   
1        Graduation   51478.705996  46.938462          3.856410   
3            Master   52145.037977  45.296296          4.240741   

   store_purchases  
2         4.607143  
0         2.833333  
4         6.537500  
1         5.912821  
3         5.981481

Does educational level and martial status has any correlation with recency? Find out.

In [80]:
df_1 = df[['marital_status','educational_level','recency', 'total_purchase']].set_index(['marital_status','educational_level'])
df_1

recency  total_purchase
marital_status educational_level                         
Single         Graduation              58              12
Married        Graduation              38               3
Divorced       Graduation              26              18
NaN            NaN                     26               6
Widowed        PhD                     94              11
...                                   ...             ...
Divorced       PhD                     85              15
               Master                  88              16
               Graduation              22               5
Relationship   PhD                     83               8
               PhD                     63              17

[500 rows x 2 columns]

In [81]:
df_1.groupby(level=['marital_status','educational_level'])['recency'].mean()

marital_status  educational_level
Divorced        Basic                62.500000
                Graduation           52.027778
                High School          87.500000
                Master               59.800000
                PhD                  40.352941
Married         Graduation           41.590909
                High School          67.000000
                Master               45.600000
                PhD                  65.000000
Relationship    Basic                39.333333
                Graduation           49.641026
                High School          44.909091
                Master               32.300000
                PhD                  41.736842
Single          Basic                52.000000
                Graduation           45.117647
                High School          51.500000
                Master               50.583333
                PhD                  49.250000
Widowed         Graduation           48.520000
                High Schoo

In [82]:
grp_1 = df_1.groupby(level=['marital_status','educational_level'])['recency'].agg([np.mean, np.max, np.min]).reset_index()
#grp_1.columns
grp_1.sort_values(by=['amax', 'amin'], ascending=False)

marital_status educational_level       mean  amax  amin
2        Divorced       High School  87.500000    99    76
19        Widowed        Graduation  48.520000    99     4
1        Divorced        Graduation  52.027778    99     3
15         Single        Graduation  45.117647    99     2
8         Married               PhD  65.000000    98    17
5         Married        Graduation  41.590909    97     1
6         Married       High School  67.000000    96     0
10   Relationship        Graduation  49.641026    96     0
13   Relationship               PhD  41.736842    96     0
0        Divorced             Basic  62.500000    94    31
22        Widowed               PhD  48.142857    94     1
16         Single       High School  51.500000    93    25
11   Relationship       High School  44.909091    92     2
3        Divorced            Master  59.800000    92     0
17         Single            Master  50.583333    91     1
18         Single               PhD  49.250000    90     6
4        Divorced               PhD  40.352941    88     3
12   Relationship            Master  32.300000    87     4
7         Married            Master  45.600000    85     4
21        Widowed            Master  33.428571    81    12
20        Widowed       High School  52.000000    80    24
9    Relationship             Basic  39.333333    59    13
14         Single             Basic  52.000000    52    52

Customer with High School and Divorced status has very good recency.

Now, let's compare  average recency and gross total_purchases.

In [85]:
grp_2 = df_1.groupby(level=['marital_status','educational_level']).agg({'recency' : 'mean', 'total_purchase' : 'sum'})
grp_2.sort_values(by=['total_purchase'], ascending=False)

recency  total_purchase
marital_status educational_level                           
Single         Graduation         45.117647             523
Married        Graduation         41.590909             421
Divorced       Graduation         52.027778             367
Relationship   Graduation         49.641026             346
               PhD                41.736842             250
Widowed        Graduation         48.520000             248
Divorced       PhD                40.352941             194
Married        PhD                65.000000             182
Widowed        PhD                48.142857             149
Married        Master             45.600000             142
Single         Master             50.583333             135
               PhD                49.250000             119
Divorced       Master             59.800000             111
Relationship   Master             32.300000              99
               High School        44.909091              90
Married        High School        67.000000              68
Widowed        Master             33.428571              65
Single         High School        51.500000              45
Relationship   Basic              39.333333              15
Divorced       High School        87.500000              10
Widowed        High School        52.000000              10
Divorced       Basic              62.500000               8
Single         Basic              52.000000               3

In [87]:
# Apply pivot table to find the aggregated sum of purchases and mean of recency per education and marital status group
pivot_table = pd.DataFrame(pd.pivot_table(df, values=['total_purchase', 'recency'], index=['marital_status'],
                        columns=['educational_level'], aggfunc={'recency': np.mean, 'total_purchase': np.sum}, fill_value=0)).reset_index()
pivot_table

marital_status    recency                                    \
educational_level                     Basic Graduation High School     Master   
0                       Divorced  62.500000  52.027778   87.500000  59.800000   
1                        Married   0.000000  41.590909   67.000000  45.600000   
2                   Relationship  39.333333  49.641026   44.909091  32.300000   
3                         Single  52.000000  45.117647   51.500000  50.583333   
4                        Widowed   0.000000  48.520000   52.000000  33.428571   

                             total_purchase                                     
educational_level        PhD          Basic Graduation High School Master  PhD  
0                  40.352941              8        367          10    111  194  
1                  65.000000              0        421          68    142  182  
2                  41.736842             15        346          90     99  250  
3                  49.250000              3        523          45    135  119  
4                  48.142857              0        248          10     65  149

Recommendations
Now after we completed the process of data cleaning and performing operations and aggregations on our dataset, we can conclude with some interesting insights about our customer base:

PhD people have the highest income, number of online and store purchases; however, High School graduates people have the highest recency or number of days since their last purchase.

Basic people account for the lowest number of web and store purchases. Married people with Graduation level have the highest total purchases. Therefore, the business recommendations for any potential marketing campaign should focus on attracting and retaining PhD people and married couples with Graduation level, and more products should be offered to satisfy the needs and interests of other categories such as Basic people that have the lowest purchases as well as High School people who are not purchasing regularly.

Additionally, further work should be conducted to understand customer behaviors and interests, for example, performing RFM analysis or regression modeling will be beneficial to study the effect of variables on the number of purchases by educational or marital status group.